### PR TITLE
Parse error constructor expression

### LINKED ISSUE #23024: Add error-constructor-expr parsing
**Description:**
$subject.
Please refer **Functional constructor** in https://htmlpreview.github.io/?https://raw.githubusercontent.com/ballerina-platform/ballerina-spec/master/lang/spec.html

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
$subject

Fixes #23024 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -3463,6 +3465,28 @@ private STNode parseFuncCall(STNode identifier) {
...
+        STNode closeParen = parseCloseParenthesis();
+        return STNodeFactory.createErrorConstructorExpressionNode(errorKeyword, openParen, args, closeParen);
+    }
+
+    /**
+     * Parse error keyword.
+     *
+     * @return Error keyword
+     */
+    private STNode parseErrorKeyword() {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3486]: I think theres a method with the similar method already

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -3463,6 +3465,28 @@ private STNode parseFuncCall(STNode identifier) {
         return STNodeFactory.createFunctionCallExpressionNode(identifier, openParen, args, closeParen);
     }
 
+    /**
+     * Parse error constructor expression.
+     *
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3470]: Shall we add the grammar rule here as a doc comment? You can check other expressions for a sample.

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 3]
```diff
@@ -698,6 +698,30 @@
                 }
             ]
         },
+        {
+            "name": "ErrorConstructorExpressionNode",
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:702]: Shall we rename this as FunctionalConstructorExpression?

File: compiler/ballerina-parser/src/test/resources/expressions/error-constructor-expr/error_constructor_expr_source_01.bal [hunk 4]
```diff
@@ -0,0 +1,3 @@
+public function foo() {
+    var er = error("ERROR_REASON", message = "error message");
+}
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/resources/expressions/error-constructor-expr/error_constructor_expr_source_01.bal:3]: Add new line

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                "/error_constructor_expr_assert_04.json");
+    }
+
+    //Invalid syntax
+    @Test
+    public void testWrongAssignment() {
+        testFile("error-constructor-expr/error_constructor_expr_source_02.bal", "error-constructor-expr" +
+                "/error_constructor_expr_assert_05.json");
+    }
+}
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java:59]: Shall we add some more tests, such as missing open-parenthesis, missing close-parenthesis, extra tokens before/after error keyword, etc

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Removed

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3479]: Done

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json (hunk 3 above)
Comment (KavinduZoysa) [reply compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json]: Done

File: compiler/ballerina-parser/src/test/resources/expressions/error-constructor-expr/error_constructor_expr_source_01.bal (hunk 4 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/test/resources/expressions/error-constructor-expr/error_constructor_expr_source_01.bal]: Done

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java (hunk 5 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java:83]: Done

**Comment by KavinduZoysa** [comment]
> Need to add changes to the error recovery.

Added the error recovery part.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 6]
```diff
@@ -3470,6 +3472,28 @@ private STNode parseFuncCall(STNode identifier) {
...
 
+    /**
+     * Parse error constructor expression.
+     * <code>functional-constructor-expr := functionally-constructible-type-reference ( arg-list )
+     * functionally-constructible-type-reference := error | type-reference</code>
+     *
+     * @param kind Syntax kind
+     * @return Functional constructor expression
+     */
+    private STNode parseFunctionalConstructorExpr(SyntaxKind kind) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3483]: Do we need the syntaxKind as a parameter for this? This method is used only in one place

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 7]
```diff
@@ -1346,6 +1348,7 @@ protected ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLo
             case KEY_SPECIFIER:
             case ERROR_TYPE_DESCRIPTOR:
             case LET_VAR_DECL:
+            case ERROR_CONSTRUCTOR_EXPRESSION:
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1351]: Even-though a context is started here, its never ended.

Should be ended after ending the args-list context.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 8]
```diff
@@ -3470,6 +3472,28 @@ private STNode parseFuncCall(STNode identifier) {
         return STNodeFactory.createFunctionCallExpressionNode(identifier, openParen, args, closeParen);
     }
 
+    /**
+     * Parse error constructor expression.
+     * <code>functional-constructor-expr := functionally-constructible-type-reference ( arg-list )
+     * functionally-constructible-type-reference := error | type-reference</code>
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3478]: Need to add proper line-breaks/paragraph-starts, to render properly. Check other doc comments.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 6 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3483]: At this moment we may have only one SyntaxKind for functional-constructor-expr. But with the implementation of other type references, new SyntaxKinds can be introduced.

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3483]: We can update the method when we have the requirement. Reason is, if anyone implementing the other type-reference decided not to use this method, then this change would become unnecessary.

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3483]: We should also look at the possibility of using the function-call node to represent error-constructor.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 8 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Added

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 7 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java]: Removed this context.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 6 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Removed the syntax kind and it was handled same as FunctionCall.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    public void testWithArgs() {
+        test("error(\"ERROR_REASON\", message = \"error message\")", "functional-constructor-expr/error-constructor" +
+                "-expr/error_constructor_expr_with_args_01.json");
+        test("error(\"ERROR_REASON\", \"error message\");", "functional-constructor-expr/error-constructor-expr" +
+                "/error_constructor_expr_with_args_02.json");
+    }
+
+    @Test
+    public void testWithAssignment() {
+        testFile("functional-constructor-expr/error-constructor-expr/error_constructor_expr_source_01.bal",
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java:45]: Don't need two levels of directories `functional-constructor-expr/error-constructor-expr/`

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 10]
```diff
@@ -3776,6 +3778,22 @@ private STNode parseFuncCall(STNode identifier) {
...
+     * Parse error constructor expression.
+     * </p>
+     * <code>
+     * functional-constructor-expr := functionally-constructible-type-reference ( arg-list )
+     * functionally-constructible-type-reference := error | type-reference
+     * </code>
+     *
+     * @return Functional constructor expression
+     */
+    private STNode parseFunctionalConstructorExpr() {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Changed

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 11]
```diff
@@ -3776,6 +3778,22 @@ private STNode parseFuncCall(STNode identifier) {
...
+        STNode functionallyConstructibleTypeReference = parseErrorKeyWord();
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Changed

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java (hunk 9 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/ErrorConstructorExprTest.java]: Changed

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 12]
```diff
@@ -1865,8 +1866,9 @@ protected ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLo
             case ERROR_KEYWORD:
                 if (isInTypeDescContext()) {
                     return ParserRuleContext.ERROR_TYPE_PARAM_START;
+                } else {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java]: Changed

**Pull request merged**